api/v1alpha5/openstackcluster_types.go

```diff
@@ -123,6 +123,14 @@ type OpenStackClusterSpec struct {
 	// +optional
 	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
 
+	// IgnoreAvailabilityZones disables the use of availability zones/failure domains,
+	// allowing Nova to schedule machines in any compatible AZ.
+	// This only affects control plane nodes - explicit failure domains can still be
+	// specified for worker nodes using MachineDeployment.spec.template.spec.failureDomain
+	// if desired.
+	// +optional
+	IgnoreAvailabilityZones bool `json:"ignoreAvailabilityZones"`
```

Comment:
can we add a section in doc talk about this usage?

Reply:
I can add a section to the docs. We have a 4 day weekend in the UK so will look at this next week.

Reply:
@jichenjc 

How would you suggest changing the docs? At the moment, availability zones are mentioned under `Configuration > Required configuration > Availability zone` which states that `OPENSTACK_FAILURE_DOMAIN` is required when using the template.

I had not intended to change this behaviour for the template, unless you think that we should? But we should probably mention that availability zones are not actually required anymore.

Reply:
yes, as you pointed out
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/docs/book/src/clusteropenstack/configuration.md#availability-zone

is the thing we are talking about , per this PR, I think it should be updated to reflect it's optional now
and may need give some example as how to disable the AZ setting..

Reply:
@jichenjc 

Should I move it out of the required config section though, since it is still required by the template? Or just mention in that section that it can be disabled if you create your own resources directly?

Reply:
I assume you talk about `ControlPlaneAvailabilityZones ` ? if so 
I think we can remove from required as it's optional if IgnoreAvailabilityZones  is true...

Reply:
I will do.

Should we remove `OPENSTACK_FAILURE_DOMAIN` from the template, since it is no longer required? @apricote @mdbooth?

Reply:
Looking at this, given that `ControlPlaneAvailabilityZones` is currently required, as @jichenjc says, this makes this a bit of an issue when this option is set. Technically you shouldn't change something from required to optional as some components relying on this API might now see an empty value which they don't know how to handle, though in this case, you can probably get away with it if you own all or most consumers of the API.

That said, if you are going to make the list optional, why not just handle the empty list case rather than having an explicit option? Bools are generally discouraged in Kube API design anyway so I'd lean towards making the `ControlPlaneAvailabilityZones` optional, adding in some context to the field description that explains what's happened when it is omitted, and then not add any extra fields

Reply:
@JoelSpeed 

Even on main, `ControlPlaneAvailabilityZones` is not required (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1alpha5/openstackcluster_types.go#L127). If not given, it takes the value of an empty list, which indicates that all the discovered AZs are suitable for control plane nodes. 

My new flag does not change this behaviour unless explicitly set to `true`. If it is not given, it defaults to `false` and the change is backwards compatible. 

However I do prefer @mdbooth’s suggestion below.